src/transformers/trainer.py

```diff
@@ -304,6 +347,14 @@ def __init__(
         # Mixed precision setup
         self.use_apex = False
         self.use_amp = False
+
+        # deepspeed manages its own fp16, so don't interfere
+        if args.deepspeed and args.fp16:
+            ds_config_file = args.deepspeed if args.deepspeed is not None else ""
+            raise ValueError(
+                f"Please use deepspeed's json configuration file {ds_config_file} to enable fp16 and not via --fp16 cl arg"
+            )
```

Comment:
Other arguments in `TrainingArguments` are gracefully handled in `_init_deepspeed` and used if nothing is set in the config file. Why have something different for FP16?

Reply:
It wasn't there in my original PR and it came from this discussion: https://github.com/huggingface/transformers/pull/9211#pullrequestreview-556207216

but that was before the whole config evolved, so yes, it should be integrated, rather then asserted - will do that next. Thank you for reviewing with fresh eyes, @sgugger

Reply:
oh, but wait - there is a reason to it. 

What about all the back-ends?

Reply:
I fixed most of it - just waiting to hear what to do with the backend.